File: android/app/src/main/java/com/insituledger/app/util/Rrule.kt
package com.insituledger.app.util

import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter

/**
 * RFC 5545 RRULE subset shared by the scheduled-transaction list, the form and
 * the materializing worker.
 *
 * Wire format: `FREQ=…[;INTERVAL=n][;UNTIL=…]`, stored verbatim in
 * scheduled_transactions.rrule. The TypeScript twin is frontend/src/lib/rrule.ts
 * and the authoritative date math is backend/internal/scheduler/scheduler.go —
 * keep the three in step.
 *
 * The one rule that must never be broken: [build] omits INTERVAL when it is 1,
 * so the six historical presets serialize byte-for-byte as they always have.
 * Clients built before v1.33.0 resolve an rrule by exact string match, and a
 * redundant `INTERVAL=1` would make them fall back to "monthly" — then
 * overwrite the real recurrence on save.
 */
enum class Freq { DAILY, WEEKLY, MONTHLY, YEARLY }

data class ParsedRrule(
    val freq: Freq,
    val interval: Int,
    /** UNTIL as "yyyy-MM-dd", or "" when the schedule has no end date. */
    val until: String
)

data class RrulePreset(
    val key: String,
    val label: String,
    val freq: Freq,
    val interval: Int
)

object Rrule {

    /** Matches the backend's validateRRule bound (helpers.go). */
    const val MAX_INTERVAL = 999
    const val CUSTOM_KEY = "custom"

    // Order is the dropdown order. Every entry must round-trip through [build]
    // to the exact string that shipped clients have always written.
    val PRESETS: List<RrulePreset> = listOf(
        RrulePreset("daily", "Daily", Freq.DAILY, 1),
        RrulePreset("weekly", "Weekly", Freq.WEEKLY, 1),
        RrulePreset("biweekly", "Biweekly", Freq.WEEKLY, 2),
        RrulePreset("monthly", "Monthly", Freq.MONTHLY, 1),
        RrulePreset("quarterly", "Quarterly", Freq.MONTHLY, 3),
        RrulePreset("yearly", "Yearly", Freq.YEARLY, 1)
    )

    private val UNITS = mapOf(
        Freq.DAILY to ("day" to "days"),
        Freq.WEEKLY to ("week" to "weeks"),
        Freq.MONTHLY to ("month" to "months"),
        Freq.YEARLY to ("year" to "years")
    )

    /**
     * Parse an rrule into its parts. Order-independent and deliberately
     * tolerant, mirroring the schedulers: unknown keys and malformed segments
     * are skipped, a missing or unrecognized FREQ falls back to MONTHLY, and a
     * non-positive or unparseable INTERVAL falls back to 1. Never throws — it is
     * given stored data that it has no way to reject.
     */
    fun parse(rrule: String): ParsedRrule {
        var freq = Freq.MONTHLY
        var interval = 1
        var until = ""

        for (part in rrule.split(";")) {
            val kv = part.split("=", limit = 2)
            if (kv.size != 2) continue
            when (kv[0]) {
                "FREQ" -> Freq.entries.find { it.name == kv[1] }?.let { freq = it }
                "INTERVAL" -> kv[1].toIntOrNull()?.let { if (it > 0) interval = it }
                "UNTIL" -> until = normalizeUntil(kv[1])
            }
        }

        return ParsedRrule(freq, interval, until)
    }

    /**
     * Serialize to the canonical wire form. [untilDate] is "yyyy-MM-dd" or "".
     * Key order is fixed at FREQ;INTERVAL;UNTIL and INTERVAL=1 is omitted, so
     * `build(parse(s))` returns `s` unchanged for every string a shipped client
     * has ever written.
     */
    fun build(freq: Freq, interval: Int, untilDate: String = ""): String {
        val sb = StringBuilder("FREQ=").append(freq.name)
        if (interval > 1) sb.append(";INTERVAL=").append(interval)
        if (untilDate.isNotEmpty()) {
            sb.append(";UNTIL=").append(untilDate.replace("-", "")).append("T235959Z")
        }
        return sb.toString()
    }

    /** The preset key for a freq/interval pair, or [CUSTOM_KEY] if it isn't one. */
    fun presetKey(freq: Freq, interval: Int): String =
        PRESETS.find { it.freq == freq && it.interval == interval }?.key ?: CUSTOM_KEY

    fun preset(key: String): RrulePreset? = PRESETS.find { it.key == key }

    /** Singular/plural unit noun, e.g. unitLabel(MONTHLY, 2) → "months". */
    fun unitLabel(freq: Freq, count: Int): String {
        val (one, many) = UNITS.getValue(freq)
        return if (count == 1) one else many
    }

    /**
     * Human label for a stored rrule, e.g. "Biweekly" or "Every 2 months". UNTIL
     * is ignored when preset-matching, so a schedule with an end date still
     * reads "Biweekly" instead of leaking the raw string.
     */
    fun label(rrule: String): String {
        val (freq, interval, _) = parse(rrule)
        val preset = PRESETS.find { it.freq == freq && it.interval == interval }
        if (preset != null) return preset.label
        return "Every $interval ${unitLabel(freq, interval)}"
    }

    /**
     * Add one recurrence step. java.time clamps the day of month to the target
     * month's last valid day (2026-01-31 + 1 month → 2026-02-28), which the Go
     * scheduler matches since v1.22.0 via addMonthsClamped.
     */
    fun advance(dateTime: LocalDateTime, freq: Freq, interval: Int): LocalDateTime =
        when (freq) {
            Freq.DAILY -> dateTime.plusDays(interval.toLong())
            Freq.WEEKLY -> dateTime.plusWeeks(interval.toLong())
            Freq.MONTHLY -> dateTime.plusMonths(interval.toLong())
            Freq.YEARLY -> dateTime.plusYears(interval.toLong())
        }

    /**
     * Advance a stored next_occurrence by one step of [rrule], returning the new
     * value and whether it is past the rule's UNTIL.
     *
     * Preserves the input's shape: a datetime comes back as RFC3339 with the
     * system zone's offset so the backend's TZ-aware comparison stays correct
     * after a local sync push; a date-only value stays date-only (TZ-agnostic by
     * design).
     */
    fun advanceDate(current: String, rrule: String): Pair<String, Boolean> {
        val hasTime = current.contains('T') || current.contains(' ')
        val dateTime = DateTimeUtil.parseFlexibleLocalDateTime(current)

        val (freq, interval, _) = parse(rrule)
        val next = advance(dateTime, freq, interval)

        // Re-read UNTIL from the raw string rather than the normalized parse, so
        // a time-of-day component in the rule is honoured to the second.
        val untilStr = rrule.split(";")
            .map { it.split("=", limit = 2) }
            .firstOrNull { it.size == 2 && it[0] == "UNTIL" }
            ?.get(1)
        val pastUntil = untilStr?.let { parseUntil(it) }?.let { next.isAfter(it) } ?: false

        val nextStr = if (hasTime) {
            next.atZone(ZoneId.systemDefault())
                .toOffsetDateTime()
                .format(DateTimeFormatter.ISO_OFFSET_DATE_TIME)
        } else {
            next.toLocalDate().toString()
        }
        return nextStr to pastUntil
    }

    // RFC 5545 UNTIL: typical forms are 20261231T235959Z or 20261231; we also
    // tolerate the same set the backend tolerates so round-tripping is safe.
    fun parseUntil(s: String): LocalDateTime? {
        val patterns = listOf(
            "yyyyMMdd'T'HHmmss'Z'",
            "yyyyMMdd'T'HHmmss",
            "yyyyMMdd",
            "yyyy-MM-dd'T'HH:mm:ss",
            "yyyy-MM-dd'T'HH:mm",
            "yyyy-MM-dd"
        )
        for (p in patterns) {
            try {
                val fmt = DateTimeFormatter.ofPattern(p)
                return if (p == "yyyyMMdd" || p == "yyyy-MM-dd") {
                    LocalDate.parse(s, fmt).atTime(23, 59, 59)
                } else {
                    LocalDateTime.parse(s, fmt)
                }
            } catch (_: Exception) { /* try next */ }
        }
        return null
    }

    // Normalize any UNTIL shape [parseUntil] accepts down to "yyyy-MM-dd".
    private fun normalizeUntil(raw: String): String =
        parseUntil(raw)?.toLocalDate()?.toString() ?: ""
}

File: frontend/src/lib/rrule.ts
// RFC 5545 RRULE subset shared by the scheduled-transaction list and form.
//
// Wire format: `FREQ=…[;INTERVAL=n][;UNTIL=…]`, stored verbatim in
// scheduled_transactions.rrule. The Kotlin twin is android/…/util/Rrule.kt and
// the authoritative date math is backend/internal/scheduler/scheduler.go —
// keep the three in step.
//
// The one rule that must never be broken: buildRrule omits INTERVAL when it is
// 1, so the six historical presets serialize byte-for-byte as they always have.
// Clients built before v1.22.0 resolve an rrule by exact string match, and a
// redundant `INTERVAL=1` would make them fall back to "monthly" — then
// overwrite the real recurrence on save.

export type Freq = 'DAILY' | 'WEEKLY' | 'MONTHLY' | 'YEARLY';

export type PresetKey = 'daily' | 'weekly' | 'biweekly' | 'monthly' | 'quarterly' | 'yearly';

export type FrequencyKey = PresetKey | 'custom';

export const CUSTOM_KEY = 'custom';

// Matches the backend's validateRRule bound (helpers.go).
export const MAX_INTERVAL = 999;

export interface ParsedRrule {
	freq: Freq;
	interval: number;
	/** UNTIL as 'YYYY-MM-DD', or '' when the schedule has no end date. */
	until: string;
}

export interface RrulePreset {
	key: PresetKey;
	label: string;
	freq: Freq;
	interval: number;
}

// Order is the dropdown order. Every entry must round-trip through
// buildRrule to the exact string that shipped clients have always written.
export const PRESETS: readonly RrulePreset[] = [
	{ key: 'daily', label: 'Daily', freq: 'DAILY', interval: 1 },
	{ key: 'weekly', label: 'Weekly', freq: 'WEEKLY', interval: 1 },
	{ key: 'biweekly', label: 'Biweekly', freq: 'WEEKLY', interval: 2 },
	{ key: 'monthly', label: 'Monthly', freq: 'MONTHLY', interval: 1 },
	{ key: 'quarterly', label: 'Quarterly', freq: 'MONTHLY', interval: 3 },
	{ key: 'yearly', label: 'Yearly', freq: 'YEARLY', interval: 1 }
];

const UNITS: Record<Freq, [string, string]> = {
	DAILY: ['day', 'days'],
	WEEKLY: ['week', 'weeks'],
	MONTHLY: ['month', 'months'],
	YEARLY: ['year', 'years']
};

const FREQS: Freq[] = ['DAILY', 'WEEKLY', 'MONTHLY', 'YEARLY'];

// Normalize any UNTIL shape the scheduler accepts down to 'YYYY-MM-DD':
// 20261231T235959Z, 20261231T235959, 20261231, 2026-12-31, 2026-12-31T23:59:59.
function normalizeUntil(raw: string): string {
	const core = raw.split('T')[0];
	let ymd = '';
	if (/^\d{8}$/.test(core)) {
		ymd = `${core.slice(0, 4)}-${core.slice(4, 6)}-${core.slice(6, 8)}`;
	} else if (/^\d{4}-\d{2}-\d{2}$/.test(core)) {
		ymd = core;
	}
	// Reject impossible dates (2026-13-45) so they can't reach a date input.
	// The Kotlin twin gets this from java.time refusing to parse them.
	return ymd && !isNaN(Date.parse(`${ymd}T00:00:00Z`)) ? ymd : '';
}

/**
 * Parse an rrule into its parts. Order-independent and deliberately tolerant,
 * mirroring the schedulers: unknown keys and malformed segments are skipped, a
 * missing or unrecognized FREQ falls back to MONTHLY, and a non-positive or
 * unparseable INTERVAL falls back to 1. Never throws — it is given stored data
 * that it has no way to reject.
 */
export function parseRrule(rrule: string): ParsedRrule {
	let freq: Freq = 'MONTHLY';
	let interval = 1;
	let until = '';

	for (const part of (rrule || '').split(';')) {
		const eq = part.indexOf('=');
		if (eq < 0) continue;
		const key = part.slice(0, eq);
		const value = part.slice(eq + 1);
		switch (key) {
			case 'FREQ':
				if ((FREQS as string[]).includes(value)) freq = value as Freq;
				break;
			case 'INTERVAL': {
				// parseInt, not the Go digit-scan, so 'abc' and '-2' both land on
				// 1. They differ only on values the backend now rejects outright.
				const n = parseInt(value, 10);
				if (Number.isFinite(n) && n > 0) interval = n;
				break;
			}
			case 'UNTIL':
				until = normalizeUntil(value);
				break;
		}
	}

	return { freq, interval, until };
}

/**
 * Serialize to the canonical wire form. `untilDate` is 'YYYY-MM-DD' or ''.
 * Key order is fixed at FREQ;INTERVAL;UNTIL and INTERVAL=1 is omitted, so
 * `buildRrule(...parseRrule(s))` returns `s` unchanged for every string a
 * shipped client has ever written.
 */
export function buildRrule(freq: Freq, interval: number, untilDate = ''): string {
	let out = `FREQ=${freq}`;
	if (interval > 1) out += `;INTERVAL=${interval}`;
	if (untilDate) out += `;UNTIL=${untilDate.replace(/-/g, '')}T235959Z`;
	return out;
}

/** The preset key for a freq/interval pair, or 'custom' if it isn't a preset. */
export function presetKeyFor(freq: Freq, interval: number): FrequencyKey {
	return PRESETS.find(p => p.freq === freq && p.interval === interval)?.key ?? CUSTOM_KEY;
}

/** Singular/plural unit noun, e.g. unitLabel('MONTHLY', 2) → 'months'. */
export function unitLabel(freq: Freq, count: number): string {
	const [one, many] = UNITS[freq];
	return count === 1 ? one : many;
}

/**
 * Human label for a stored rrule, e.g. 'Biweekly' or 'Every 2 months'. UNTIL is
 * ignored when preset-matching, so a schedule with an end date still reads
 * 'Biweekly' instead of leaking the raw string.
 */
export function rruleLabel(rrule: string): string {
	const { freq, interval } = parseRrule(rrule);
	const preset = PRESETS.find(p => p.freq === freq && p.interval === interval);
	if (preset) return preset.label;
	return `Every ${interval} ${unitLabel(freq, interval)}`;
}
